hazelcast/src/main/java/com/hazelcast/map/impl/operation/AddIndexOperation.java

```diff
@@ -93,6 +130,16 @@ public void runInternal() {
 
         recordStore.forEach((dataKey, record) -> {
             Object value = Records.getValueOrCachedValue(record, serializationService);
+            if (recreateMetadata) {
+                // The call below should be fairly cheap if the
+                // key/value is not an HazelcastJsonValue or
+                // Data representation of HazelcastJsonValue.
+                // This call will perform couple of instanceof
+                // checks and try to read the serialization type
+                // of the Data key/value and do nothing if it
+                // is not something we can extract metadata from.
+                metadataStore.initAndSet(dataKey, value);
```

Comment:
Do we need to re-create metadata for `dataKey` if it already exists? If no need, we can have a variant of `updateValueMetadataIfNecessary` method like `addOrUpdateMetadata`. By doing this only value metadata is created, key metadata can be used as is.